clash-cores/src/Clash/Cores/Xilinx/BlockRam/Internal.hs

```diff
@@ -0,0 +1,164 @@
+{-|
+  Copyright   :  (C) 2023 QBayLogic B.V.
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Model implementation for primitives in "Clash.Cores.Xilinx.BlockRam".
+-}
+
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Clash.Cores.Xilinx.BlockRam.Internal where
+
+import Clash.Explicit.Prelude hiding (enable)
+import Clash.Explicit.BlockRam.Model (tdpbramModel)
+import Clash.XException.MaybeX (MaybeX(..))
+
+import Clash.Annotations.Primitive (Primitive(InlineYamlPrimitive), hasBlackBox)
+import Clash.Cores.Xilinx.BlockRam.BlackBoxes (tdpbramBBF)
+import Clash.Sized.Internal.BitVector (BitVector(BV), undefined#)
+
+import Data.String.Interpolate (__i)
+import GHC.Stack (HasCallStack)
+
+import qualified Data.Sequence as Seq
+
+-- $setup
+-- >>> import Clash.Explicit.Prelude
+
+
+-- | Returns a byte enable such that it agrees with the global enable of a
+-- block RAM. I.e., a deasserted global enable will return a deasserted byte
+-- enable.
+--
+-- >>> mergeEnableByteEnable @3 (pure undefined) (pure $(bLit "01."))
+-- IsDefined_ 0b0..
+-- >>> mergeEnableByteEnable @3 (pure False) (pure $(bLit "01."))
+-- IsDefined_ 0b000
+-- >>> mergeEnableByteEnable @3 (pure True) (pure $(bLit "01."))
+-- IsDefined_ 0b01.
+-- >>> mergeEnableByteEnable @3 (pure True) (pure undefined)
+-- IsX_ ...
+--
+mergeEnableByteEnable ::
+  KnownNat n =>
+  -- | Global enable
+  MaybeX Bool ->
+  -- | Byte enable
+  MaybeX (BitVector n) ->
+  -- | Byte enable combined with global enable
+  MaybeX (BitVector n)
+mergeEnableByteEnable enaX byteEnaX =
+  case enaX of
+    IsX _           -> liftA2 (.&.) (pure undefined#) byteEnaX
+    IsDefined True  -> byteEnaX
+    IsDefined False -> pure 0
+
+-- | Clone each bit in a mask 8 times.
+--
+-- >>> byteMaskToBitMask (0b01 :: BitVector 2)
+-- 0b0000_0000_1111_1111
+--
+byteMaskToBitMask :: KnownNat n => BitVector n -> BitVector (8 * n)
+byteMaskToBitMask = pack . map go . unpack
+ where
+  go :: Bool -> BitVector 8
+  go True = maxBound
+  go False = 0
+
+-- | Determine whether a write enable is (potentially) active
+isActiveWriteEnable :: KnownNat nBytes => MaybeX (BitVector nBytes) -> MaybeX Bool
+isActiveWriteEnable = fmap (/= 0)
+
+-- | Update a true dual port block RAM with byte enables.
+updateRam ::
+  ( KnownNat nBytes
+  , BitSize a ~ (8 * nBytes)
+  , BitPack a
+  , NFDataX a ) =>
+  Int ->
+  MaybeX (BitVector nBytes) ->
+  a ->
+  Seq.Seq a ->
+  Seq.Seq a
+updateRam addr (IsX _) dat mem =
+  updateRam addr (pure undefined#) dat mem
```

Comment:
It is correct, but confusing, to construct `IsDefined undefined#`. But okay.

Reply:
I'll add a comment